File: Julia/deflationoperator.jl
function deflation_step_adjustment(x::AbstractVector{T}, update::AbstractVector{T}, known_roots::AbstractVector, M::AbstractMatrix{T}, p::Int=2, α::T=one(T)) where T
        dMy = getdMy(x, update, known_roots, M, p, α)
        minv = one(T) / deflation_evaluate(x, known_roots, M, p, α)
        one(T)/(one(T) - minv*dMy)
end

function getdMy(x::AbstractVector, update::AbstractVector, known_roots::AbstractVector, M::AbstractMatrix, p::Int=2, α::T=one(T)) where T
    deriv = deflation_derivative(x, known_roots, M, p, α)
    # defcon has a minus sign here, but that's because PETSc
    # calculates the update so that state = state - update rather
    # than state = state + update
    deriv*M*update
end

function norm_squared(y::AbstractVector, root::AbstractVector, M::AbstractMatrix)
    # Inner product matrix is necessary to compute correct norm 
    # induced by function space
    (y - root)' *  M * (y - root)
end

function derivative_norm_squared(y::AbstractVector, root::AbstractVector)
    # Derivative is in the dual space, hence should be a row vector
    2 * (y - root)'
end

function deflation_evaluate(y::AbstractVector{T}, known_roots::AbstractVector, M::AbstractMatrix{T}, p::Int=2, α::T=one(T)) where T
    m = one(T)
    for iter = 1:lastindex(known_roots)
        normsq = norm_squared(y, known_roots[iter], M)
        factor = normsq^(-p/2) + α
        m = m * factor
    end
    m
end

function deflation_derivative(y::AbstractVector{T}, known_roots::AbstractVector, M::AbstractMatrix, p::Int=2, α::T=one(T)) where T

    N = lastindex(known_roots)
    
    normsqs = [norm_squared(y, known_roots[i], M) for i in 1:N]
    dnormsqs = [derivative_norm_squared(y, known_roots[i]) for i in 1:N]

    factors = [normsqs[i]^(-p/2) + α  for i in 1:N]
    dfactors = [(-p/2) * normsqs[i]^((-p/2) - one(T)) for i in 1:N]
   
    sum((prod(factors) ./ factors) .* dfactors .* dnormsqs)
end

File: src/Python/deflationoperator.py
import numpy as np

class DeflationOperator(object):    
    def __init__(self, known_roots, M, p, a):
        self.known_roots = known_roots
        self.M = M
        self.p = p
        self.a = a

    def update_known_roots(self, known_roots):
        self.known_roots = known_roots

    def tau(self, x, update):
        dMy = self.getdMy(x, update)
        minv = 1.0 / self.deflation_evaluate(x)
        return 1.0 / (1.0 - minv*dMy)

    def getdMy(self, x, update):
        # defcon has a minus sign here, but that's because PETSc
        # calculates the update so that x = x - update rather
        # than x = x + update
        deriv = self.deflation_derivative(x)
        return deriv.dot(self.M.dot(update))

    def norm_squared(self, y, root):
        return (y-root).dot(self.M.dot(y-root))

    def derivative_norm_squared(self, y, root):
        return 2 * (y-root)

    def deflation_evaluate(self, y):
        m = 1.0
        for i in range(len(self.known_roots)):
            normsq = self.norm_squared(y, self.known_roots[i])
            factor = normsq**(-self.p/2) + self.a
            m = m * factor
        return m

    def deflation_derivative(self, x):
        N = len(self.known_roots)
        normsqs = [self.norm_squared(x, self.known_roots[i]) for i in range(N)]
        dnormsqs =  [self.derivative_norm_squared(x, self.known_roots[i]) for i in range(N)]

        factors = [normsqs[i]**(-self.p/2) + self.a for i in range(N)]
        dfactors = [(-self.p/2) * normsqs[i]**((-self.p/2) - 1.0) for i in range(N)]

        eta = np.prod(factors)
        deta = np.sum([eta/factors[i] * dfactors[i] * dnormsqs[i] for i in range(N)], axis=0)
        return deta        

    def newtonls(self, x, residual, jacobian, tol=1e-9, max_iter=1000):
        M = self.M
        known_roots = self.known_roots
        p = self.p
        a = self.a

        r = residual(x)
        J = jacobian(x)

        norm_residual = np.linalg.norm(r)
        print("Iteration 0, residual norm = %s\n" %norm_residual)

        i = 0
        while norm_residual > tol and i < max_iter:
            update = -np.linalg.solve(J, r)   
            if known_roots:
                tau = self.tau(x, update)
                update = tau * update
            # update = nls.LineSearch.adjust(x, update, nls.damping)
            x = x + update
            r = residual(x)
            J = jacobian(x)
            norm_residual = np.linalg.norm(r)
            i += 1
            print("Iteration %s, residual norm = %s\n" %(i, norm_residual))

        if iter == max_iter:
            print("Iteration max reached")

        return x